common/src/main/java/org/conscrypt/ConscryptEngine.java

```diff
@@ -1373,23 +1354,41 @@ public void onSSLStateChange(int type, int val) {
                 case SSL_CB_HANDSHAKE_START: {
                     // For clients, this will allow the NEED_UNWRAP status to be
                     // returned.
-                    engineState = STATE_HANDSHAKE_STARTED;
+                    state = STATE_HANDSHAKE_STARTED;
                     break;
                 }
                 case SSL_CB_HANDSHAKE_DONE: {
-                    if (engineState != STATE_HANDSHAKE_STARTED
-                            && engineState != STATE_READY_HANDSHAKE_CUT_THROUGH) {
+                    if (state != STATE_HANDSHAKE_STARTED
+                            && state != STATE_READY_HANDSHAKE_CUT_THROUGH) {
                         throw new IllegalStateException(
-                                "Completed handshake while in mode " + engineState);
+                                "Completed handshake while in mode " + state);
                     }
-                    engineState = STATE_HANDSHAKE_COMPLETED;
+                    state = STATE_HANDSHAKE_COMPLETED;
                     break;
                 }
             }
         }
     }
 
     @Override
+    public boolean onNewSessionEstablished(long sslSessionNativePtr) {
+        try {
+            // Cache the newly established session.
+            AbstractSessionContext ctx = sessionContext();
+            ctx.cacheSession(SslSessionWrapper.newInstance(sslSessionNativePtr, sslSession));
```

Comment:
`SslSessionWrapper.newInstance` takes ownership of `sslSessionNativePtr`, right? So suppose `ctx.cacheSession` throws an exception. We've already taken ownership of `sslSessionNativePtr`, but this function will return false. That will cause the C code to think we *haven't* taken ownership and double-free.

Two possible fixes:

* The C-level callback returns true if it's called the callback at all and this method returns `void`. The promise would be that the Java code, no matter what happens, takes ownership of `sslSessionNativePtr`. That would mean you want to wrap `sslSessionNativePtr` as the first line I guess.

* Alternatively, `SslSessionWrapper.newInstance` does not take ownership. Rather it calls `SSL_SESSION_up_ref` before stashing the pointer on itself. The Java `onNewSessionEstablished` then also does not take ownership of the provided reference, so the C-level callback always returns false. As in the first option, this function also now returns `void`.

(It's perfectly fine for the return value to not match whether the session was. It's some bizarre OpenSSL legacy thing and doesn't terminate the connection or anything. Think of the return value as that goofy `out_copy` parameter to the other callback. It's another one of those weird things where OpenSSL should have just picked one and left it at that. Pick whichever return value works better and don't think too hard about making it match whether operations succeeded or not.)